ahcer/src/app/episode-report/episode-report.component.ts

```diff
@@ -0,0 +1,283 @@
+import {Episode} from "../models/episode";
+import {Patient} from "../models/patient";
+import {EpisodeService} from "../services/episode.service";
+import {PatientServices} from "../services/patient.service";
+import {UsersService} from "../services/users.service";
+import {finalize} from "rxjs";
+import firebase from "firebase/compat";
+import Timestamp = firebase.firestore.Timestamp;
+import {MedicationService} from "../services/medication.service";
+import {MatTableDataSource} from "@angular/material/table";
+import {MatSort} from "@angular/material/sort";
+import {MatPaginator} from "@angular/material/paginator";
+
+@Component({
+  selector: 'app-episode-report',
+  templateUrl: './episode-report.component.html',
+  styleUrls: ['./episode-report.component.scss']
+})
+export class EpisodeReportComponent implements OnInit, AfterViewInit {
+
+  displayedColumns: string[] = ['startTime', 'endTime', 'duration', 'symptoms',
+                                'rescueMeds', 'prescriptionMeds', 'triggers'];
+  loadingPatient: boolean = false;
+  loadingEpisodes: boolean = false;
+  loadingRescueMeds: boolean = false;
+  episodes: Episode[]=[];
+  patients: Patient[];
+  rescueMedNames: Object = {};
+  currentPatient : Patient;
+  episodes_count: number;
+  dataSource: MatTableDataSource<Episode> = new MatTableDataSource<Episode>();
+
+  constructor(private medicationService: MedicationService,
+              private episodeService: EpisodeService,
+              private patientsService: PatientServices,
+              private usersService: UsersService) { }
+
+  ngOnInit(): void {
+    this.loadingPatient = true;
+
+    this.patientsService.getPatients().subscribe(
+      patients => {this.patients = patients}
+    )
+    this.usersService.getLastViewedPatient().subscribe(
+      (patientId)  => {
+        if(patientId) {
+          this.loadPatient(patientId);
+          this.loadEpisodes(patientId);
+          this.loadRescueMeds(patientId);
+        }
+        else {
+          this.loadingPatient = false;
+        }
+      })
+  }
+
+  @ViewChild(MatSort) sort: MatSort;
+  @ViewChild(MatPaginator) paginator: MatPaginator;
+
+  sortingDataAccessor = (item: Episode, property: string) => {
+    switch (property) {
+      case 'startTime':
+        return item.startTime.seconds;
+      case 'endTime':
+        return item.endTime?.seconds;
+      case 'duration':
+        return (item.endTime)? item.endTime.seconds - item.startTime.seconds: null;
+      case 'symptoms':
+        return this.displaySymptomsString(item.symptoms);
+      default:
+        return item[property];
+    }
+  }
+
+  reloadDataSource() {
+    this.dataSource.data = this.episodes;
+    this.dataSource.sort = this.sort;
+    this.dataSource.sortingDataAccessor = this.sortingDataAccessor;
+    this.dataSource.paginator = this.paginator;
+  }
+
+  @HostListener('window:beforeprint',['$event'])
+  onBeforePrint(event){
```

Comment:
Is this event variable necessary? We aren't using it.

Reply:
I just removed it!